File: server/src/main/java/ru/fullrest/mfr/server/model/service/NotificationService.java
package ru.fullrest.mfr.server.model.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.telegram.telegrambots.bots.DefaultAbsSender;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;
import ru.fullrest.mfr.server.config.PropertiesConfiguration;

import java.util.concurrent.ConcurrentLinkedQueue;


@Service
@Log4j2
@RequiredArgsConstructor
public class NotificationService {

    private final PropertiesConfiguration configuration;

    private final DefaultAbsSender telegramBot;

    private ConcurrentLinkedQueue<String> unsentMessages = new ConcurrentLinkedQueue<>();


    public void sendNotification(String text) {
        SendMessage message = new SendMessage();
        message.setChatId(configuration.getChatId());
        message.setText(text);
        try {
            telegramBot.execute(message);
        } catch (TelegramApiException e) {
            log.error(e);
            unsentMessages.offer(text);
        }
    }

    @Scheduled(fixedDelay = 120_000L, initialDelay = 180_000L)
    private void checkUnsentMessages() {
        for (int i = 0; i < unsentMessages.size(); i++) {
            String message = unsentMessages.poll();
            if (message != null) {
                sendNotification(message);
            }
        }
    }
}

File: server/src/main/kotlin/com/lezenford/mfr/server/service/NotificationService.kt
package com.lezenford.mfr.server.service

import com.lezenford.mfr.server.configuration.properties.ApplicationProperties
import com.lezenford.mfr.server.event.SendMessageEvent
import com.lezenford.mfr.server.model.entity.TelegramUser
import com.lezenford.mfr.server.service.model.TelegramUserService
import org.springframework.boot.context.event.ApplicationReadyEvent
import org.springframework.context.ApplicationEventPublisher
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Service
import org.telegram.telegrambots.meta.api.methods.send.SendMessage
import javax.annotation.PreDestroy

@Service
class NotificationService(
    private val applicationEventPublisher: ApplicationEventPublisher,
    private val telegramUserService: TelegramUserService,
    private val applicationProperties: ApplicationProperties
) {
    private var successfullyStarted = false

    @EventListener(ApplicationReadyEvent::class)
    fun successfullyStarted() {
        telegramUserService.findAll().filter { it.role == TelegramUser.Role.ADMIN }.forEach {
            applicationEventPublisher.publishEvent(
                SendMessageEvent(
                    this,
                    SendMessage(it.id.toString(), "Сервер успешно запущен. Версия: ${applicationProperties.version}")
                )
            )
        }
        successfullyStarted = true
    }

    @PreDestroy
    fun preDestroy() {
        if (successfullyStarted) {
            telegramUserService.findAll().filter { it.role == TelegramUser.Role.ADMIN }.forEach {
                applicationEventPublisher.publishEvent(
                    SendMessageEvent(this, SendMessage(it.id.toString(), "Сервер остановлен"))
                )
            }
        }
    }
}